### ISSUE TITLE
Overriding default instance type for beanstalk environment is not working

### ISSUE THREAD

**Issue by puneetpandey** [aws/aws-cli#6013]
I have a cloudformation template and I would like to choose instance type dynamically for each environment. Here's what I am trying to do:

````
AWSTemplateFormatVersion: 2010-09-09
Description: MyApp CloudFormation
Conditions:
  ProductionCondition: !Equals [!Ref 'AWS::StackName', production]
  NotProductionCondition: !Not [!Equals [!Ref 'AWS::StackName', production]]
Parameters:
  RailsApplicationTemplate:
    Type: AWS::ElasticBeanstalk::ConfigurationTemplate
    Properties:
      OptionSettings:
      - Namespace: aws:autoscaling:launchconfiguration
        OptionName: InstanceType
        Value: t2.micro
      - Namespace: aws:elasticbeanstalk:application:environment
        OptionName: AWS_ENV
        Value: web
  RailsWebEnvironment:
    Type: AWS::ElasticBeanstalk::Environment
    DependsOn:
      - RailsApplicationTemplate
    Properties:
      OptionSettings:
      - Namespace: aws:autoscaling:launchconfiguration
        OptionName: InstanceType
        Value: !If [ProductionCondition, 't2.large', 't2.medium']
  RailsWorkerEnvironment:
    Type: AWS::ElasticBeanstalk::Environment
    DependsOn:
      - RailsApplicationTemplate
    Properties:
      OptionSettings:
      - Namespace: aws:autoscaling:launchconfiguration
        OptionName: InstanceType
        Value: t2.medium
      - Namespace: aws:elasticbeanstalk:application:environment
        OptionName: AWS_ENV
        Value: worker
````

Whenever I run ` aws cloudformation update-stack --stack-name staging --template-body file://cloudformation/template.yml --capabilities CAPABILITY_IAM` the CF command does not change the `InstanceType` and always use `t2.micro` as its value. On the other side, it does change the default value of `AWS_ENV` to `Worker` for Worker environment.

Other solutions I have tried are:
1. By removing `InstanceType` from `RailsApplicationTemplate` and use it individually for each environment. - But that didn't work.

Is this supported? Am I doing something wrong here?
Please advise.

**Comment by stobrien89** [comment]
Hi @puneetpandey,

Thanks for writing in! It's tough to say whether or not this should work without knowing the condition of the stack at the time the update is being attempted— is a t2.micro instance running at the time of the attempted update? Changing an `AWS::ElasticBeanstalk::Environment` `OptionSetting` `Value` requires "no interruption", per the [docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-beanstalk-option-settings.html) and changing an instance type would cause an interruption.  

If this isn't the case, we can definitely take a look at the debug logs to get a better idea of what's happening here!

**Comment by puneetpandey** [comment]
@stobrien89 thanks for replying!

After executing the CF command, when I checked the events for the environment via ElasticBeanstalk from AWS console and below is the trail:

````
2021-03-12 13:50:05 UTC+0530 | INFO | Environment update completed successfully.
-- | -- | --
2021-03-12 13:50:05 UTC+0530 | INFO | Successfully deployed new configuration to environment.
2021-03-12 13:50:05 UTC+0530 | INFO | New application version was deployed to running EC2 instances.
2021-03-12 13:49:44 UTC+0530 | INFO | Deploying new version to instance(s).
2021-03-12 13:49:09 UTC+0530 | INFO | Updating environment stageWorker's configuration settings.
2021-03-12 13:48:57 UTC+0530 | INFO | Environment update is starting.
````
Basically, it never updated the configuration setting and during the configuration update EC2 instance remain uninterrupted means neither it rebooted or restarted. And no downtime as well. So I guess something didn't work while the configuration was in progress.

This only works if I change instance type to any value inside `RailsApplicationTemplate`. So perhaps overriding instance type is ignored in my case. And I think raising this to AWS level makes more sense because I don't see anything weird with `aws` command. But if you can point this to the right direction that would be really helpful.

**Comment by stobrien89** [comment]
Hi @puneetpandey,

Thank you for the additional information.  My apologies for any confusion, but what I should have said in my last comment is that attempting to override the instance type (of a running instance) will be ignored in any case because changing an instance type requires an interruption, which is not allowed to happen in order for a `AWS::ElasticBeanstalk::Environment` `OptionSetting` value to update.  

In other words, it appears this is expected behavior, but please let me know if I missed anything!  This may be a case where the instance types need to be updated manually via `delete-stack-instances` and `create-stack-instances`.

**Comment by puneetpandey** [comment]
That totally make sense to me @stobrien89 and thanks for clarifying. I will try a different solution approach within `AWS::ElasticBeanstalk::ConfigurationTemplate`

**Comment by stobrien89** [comment]
hi @puneetpandey,

Of course— please don't hesitate to reach out if you need any additional support!
